pkg/logger/mock.go

```diff
@@ -0,0 +1,125 @@
+	Message string
+	// Fields holds each field added to this logged event. Some logged data is
+	// translated into this map array:
+	//
+	// 	Event.WithScope("foo")   => MockLog.Fields["scope"] = "foo"
+	// 	Event.WithError(someErr) => MockLog.Fields["error"] = someErr
+	// 	Event.WithCaller("foo", 42)
+	// 		=> MockLog.Fields["caller"] = "foo"
+	// 		=> MockLog.Fields["line"] = 42
+	Fields map[string]interface{}
+	// FieldNames is a slice of strings with all the keys added to the Fields
+	// map. This includes the custom mapping of Event.WithScope,
+	// Event.WithError, and Event.WithCaller as mentioned in the Fields docs.
+	FieldNames []string
+}
+
+// NewMock creates a new Logger interface compatible type that holds additional
+// fields of the logs that has been submitted.
+func NewMock() *MockLogger {
+	return &MockLogger{}
+}
+
+// Debug creates a new event using new contexts connected to this mock logger of
+// "debugging" logging level or higher.
+func (log *MockLogger) Debug() Event { return log.newEvent(LevelDebug) }
+
+// Info creates a new event using new contexts connected to this mock logger of
+// "information" logging level or higher.
+func (log *MockLogger) Info() Event { return log.newEvent(LevelInfo) }
+
+// Warn creates a new event using new contexts connected to this mock logger of
+// "warning" logging level or higher.
+func (log *MockLogger) Warn() Event { return log.newEvent(LevelWarn) }
+
+// Error creates a new event using new contexts connected to this mock logger of
+// "error" logging level or higher.
+func (log *MockLogger) Error() Event { return log.newEvent(LevelError) }
+
+// Panic creates a new event using new contexts connected to this mock logger of
+// "panic" logging level or higher.
+//
+// Compared to the other logging events, after submitting the logged
+// messages this method calls panic with the final message string.
+func (log *MockLogger) Panic() Event { return log.newEvent(LevelPanic) }
+
+func (log *MockLogger) newEvent(level Level) Event {
+	var done DoneFunc
+	if level == LevelPanic {
+		done = panicString
+	}
+	return event{
+		level: level,
+		ctxs:  []Context{newMockContext(log)},
+		done:  done,
+	}
+}
+
+func newMockContext(logger *MockLogger) Context {
+	return mockCtx{
+		MockLog: MockLog{
+			Fields: map[string]interface{}{},
+		},
+		logger: logger,
+	}
+}
+
+type mockCtx struct {
+	MockLog
+	logger *MockLogger
+}
+
+func (c mockCtx) WriteOut(level Level, message string) {
+	c.Level = level
+	c.Message = message
+	c.logger.Logs = append(c.logger.Logs, c.MockLog)
+}
+
+func (c mockCtx) SetCaller(file string, line int) Context {
+	c.Fields["caller"] = file
+	c.Fields["line"] = line
+	c.FieldNames = append(c.FieldNames, "caller", "line")
+	return c
+}
+
+func (c mockCtx) SetScope(v string) Context                        { return c.addField("scope", v) }
+func (c mockCtx) SetError(v error) Context                         { return c.addField("error", v) }
```

Comment:
Implementation here doesn't seem to match the documentation for SetCaller, SetScope, and SetError.
Not sure if it matters here though.

Reply:
Quite perceptive. It should, and I think it matters. Will fix!

Reply:
The implementation did set the fields in the map.

I conclude that this MockLogger works as expected. I've renamed the field `c.FieldNames` to `c.FieldsAdded` to maybe be more descriptive.